[IMP] project: implement new project task state

bastvdn opened the pull request:
# Specs Mark as Done

## Overview

Before this PR the task state was fixed by the kanban_state field which was useful when you use the stage of the project as parts of a pipeline,
but not relevant when users are using stages as bucket lists. (specific examples at the end of the specs)

The goal of this PR is to provide users a way to mark their tasks as done with a simple button press,
while keeping the option to label a task as Approved, Canceled or Requesting changes like in the old kanban_state field.

The kanban_state of a task had no impact whatsoever on other tasks of the pipe, we would like to change that and make the task state have an influence on its dependent tasks.

The state will also have influence over the 'recurrent' tasks (to be implemented in Task #3084945)

If you want a better description of those changes with screenshot and colors check specs of:

[Task-3084930](https://www.odoo.com/web#id=3084930&cids=1&menu_id=4720&action=333&active_id=4105&model=project.task&view_type=form)

PRs:

See https://github.com/odoo/enterprise/pull/35359
See https://github.com/odoo/upgrade/pull/4367

-----------------------------------------

## Interaction with blocking tasks:
the closed values which mark the task as closed or finished:
- Done
- Canceled

The Open values when the task isn't finished yet:
- In progress
- Changes Requested
- Approved
- Waiting (which is not selectable)

## Where to change the state of a task:
- For kanban and form views: same place as kanban_state (bottom right of kanban card, top right of form view)
- For list view:  left of list (after task priority)
more details about the state widget in state field widgets part

## Interaction with existing fields
- is_closed: which was determined by the task.stage_id.fold, now a task is closed when in one of the following stages
  - Done
  - Canceled
a closed task is considered as finished, the time of the closing will be stored in the date_last_stage_update field

- is_blocked: a task is considered blocked if ANY of its blocking task is in one of the blocking states (more details about this in the following part Interaction with blocking tasks):
  - in Progress
  - Changes Requested
  - Approved
  - Waiting

!! important !! is_closed and is_blocked are not mutually exclusive, you can have a task that blocked and is closed at the same time, the reason why will be explained late

date_last_stage_update: this field is updated everytime the task goes into a closing state OR when the task changes stage.
We need to check that the value is updated in each case (using the already available filter)

## Interaction with blocking tasks
the state of a task can now be changed by its blocking tasks following the logic:

if ANY of the blocking tasks is NOT closed (so its state is in one of the open values) the task is considered as blocked

- if a task is blocked and NOT closed its state will switch to Waiting
  - the Waiting state will display an unclickable hourglass icon on the task kanban/list views, once in the waiting state you can't change the state of the taskfrom the kanban/list views
  - a blocked task state can be changed through the form view, so you can override the 'block' by choosing a closed state (only done or canceled)
     - once overriden, the task will change to the closed state the user wants, but the task is still blocked so in case where the user comes back to an open state, the task will automatically switch back to the waiting state (according to the state before the block)
- if the blocking task switches to a non-blocking state, the task will not be considered as blocked anymore and its state will switch back to In Progress

![image](https://user-images.githubusercontent.com/36259462/223415487-d59e26c2-7c21-4240-8b69-ae4307724363.png)


## Default values
the default value is always in progress

## Special cases
when a task is moved from a stage to another one
- if the state was in one of the open states (approved, changes requested, in progress ) the state goes back to In Progress
- if not, the state stays the same

when a task is moved from a project to another one
- the state goes back to In Progress
when a task is duplicated
- if the state was in one of the open states (approved, changes requested, in progress ) the state goes back to In Progress
- if not, the state stays the same

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/107593).

xavierbol commented on addons/project/models/project.py:
```diff
@@ -62,8 +64,29 @@
     'child_ids',
     'parent_id',
     'priority',
+    'state',
+    'state_approval_mode'
```
if those fields are writable in project sharing then they are also readable. 🙂 
That is, no need to add those fields in the readable fields. 🙂 

https://github.com/odoo/odoo/blob/b9f0963ce3ee318a87aa011bc1f0c7b2a5013169/addons/project/models/project.py#L1347-L1349

adr-odoo commented:
Feel free to ping me when this PR is ready.

MissingNoShiny reviewed (commented):
Hello, congratulations for your work! :star_struck: 
I left some comments :slightly_smiling_face:

bastvdn commented on addons/project/tests/test_task_state.py:
```diff
@@ -0,0 +1,67 @@
+        })
+        self.assertEqual(self.task_1.state, 'in_progress', "task_1 state should automatically switch back to in_progress when its dependency closes")
+
+    def test_change_task_project(self):
+        pass
+
+    def test_duplicate_dependent_task(self):
+        pass
```
Will be done in the next fixup commit

bastvdn commented on addons/project/views/project_views.xml:
```diff
@@ -65,7 +69,7 @@
                 </field>
                 <field name="description" position="after">
-                    <filter string="My Tasks" name="my_tasks" domain="[('user_ids', 'in', uid)]"/>
+                    <filter string="My Tasks" name="my_tasks"/>
```
nope, reverted ^^'

bastvdn commented on addons/project/models/project.py:
```diff
@@ -1149,12 +1161,22 @@ def _read_group_personal_stage_type_ids(self, stages, domain, order):
+        # ('done', 'Ready'),
+        # ('blocked', 'Blocked')], string='Status_deprecated',
+        # copy=False, default='normal', required=True, compute='_compute_kanban_state', readonly=False, store=True)
+    # kanban_state_label = fields.Char(compute='_compute_kanban_state_label', string='Kanban State Label', tracking=True, task_dependency_tracking=True)
```
Yes, removing those field will be done in the next commit :+1:

bastvdn commented on addons/project/models/project.py:
```diff
@@ -1166,8 +1188,9 @@ def _read_group_personal_stage_type_ids(self, stages, domain, order):
         readonly=True,
-        help="Date on which the stage of your task has last been modified.\n"
-            "Based on this information you can identify tasks that are stalling and get statistics on the time it usually takes to move tasks from one stage to another.")
+        help="Date on which the state of your task has last been modified.\n"
```
Yep, i actually had 2 fields but it was deemed too confusing (having a _date_last_stage_update_ and a _date_last_state_update_), so we joined them in one field

bastvdn commented on addons/project/models/project.py:
```diff
@@ -23,7 +23,6 @@
-    'kanban_state_label',
     'project_id',
     'display_project_id',
     'color',
```
ok, like mentionned before, i'll also remove every reference to _legend_**_ which also makes the change of stage bug in portal view

bastvdn commented:
Thanks a lot for the review :) very insightful
I fixed the minor problem you pointed
Some of the problems will need a little bit more time but will done short-term: (hopefully done by the 24 Feb)
- write last implementation tests
- remove all ref to kanban_state_label and legend_**
- finish migration script

Here are some improvements that should be done in more long term (may take a bit more time)

- Remove all ref to is_closed and is_blocked
- Some improvements for the frontend
  - some changes on the progressbar
  - fixed width for the tree view widget
  - write Javascript tests
- Create a tour
What would you say is the more important

And i have a small question to finish, haven't you seen any problem with the access rights/groups? I did not implement this feature with the access rights in mind at all (it does not seem that important for that small of a feature but still.. may be important)

MissingNoShiny commented:
> Here are some improvements that should be done in more long term
> 
> * Remove all ref to is_closed and is_blocked
> * Some improvements for the frontend
>   
>   * some changes on the progressbar
>   * fixed width for the tree view widget
>   * write Javascript tests
> * Create a tour
> 
> What would you say is the more important?

I would say tests/tours can wait a bit if necessary, eliminating visible problems is more important (and other bugs that might arise in testing).
 `is_closed` and `is_blocked` should be removed in this PR if possible.

> And i have a small question to finish, haven't you seen any problem with the access rights/groups? I did not implement this feature with the access rights in mind at all (it does not seem that important for that small of a feature but still.. may be important)

I didn't notice anything problematic besides the portal rights. Hopefully if there are big issues, they will come up during testing :slightly_smiling_face:

xavierbol commented on addons/project/views/project_views.xml:
```diff
@@ -1448,7 +1401,7 @@
                                 <div class="o_kanban_record_body">
                                     <field name="tag_ids" widget="many2many_tags" options="{'color_field': 'color'}"/>
-                                    <div t-if="record.date_deadline.raw_value" name="date_deadline" attrs="{'invisible': [('is_closed', '=', True)]}">
+                                    <div t-if="record.date_deadline.raw_value" name="date_deadline" invisible="True" attrs="{'invisible': [('is_closed', '=', True)]}">
```
Then we can remove the div? 🤔

xavierbol reviewed (commented):
Great job! Thanks for your work, I left some comments. 🙂

bastvdn commented on addons/project/models/project.py:
```diff
@@ -1545,10 +1580,16 @@ def _compute_dependent_tasks_count(self):
+    def _inverse_partner_email(self):
+        for task in self:
+            if task.partner_id and task.partner_email != task.partner_id.email:
+                task.partner_id.email = task.partner_email
```
:+1: Indeed

bastvdn commented on addons/project/models/project.py:
```diff
@@ -1366,9 +1367,43 @@ def _search_is_private(self, operator, value):
+        if not self.allow_task_dependencies and self.state == '04_waiting_normal':
+            self.state = '01_in_progress'
+
+    @api.onchange('stage_id')
+    def _onchange_stage_id(self):
+        if self.state != '04_waiting_normal' and self.state not in CLOSED_STATES:
+            self.state = '01_in_progress'
```
Same thing as for allow task dependencies, here we are just trying to handle a change of stage,
how can you handle that in the **_compute** function?

xavierbol replied:
you can let this change, you are right. 🙂

bastvdn commented on addons/project/models/project.py:
```diff
@@ -1366,9 +1367,43 @@ def _search_is_private(self, operator, value):
+                if task.state == '04_waiting_normal':
+                    task.state = '01_in_progress'
+
+    @api.onchange('allow_task_dependencies')
+    def _onchange_allow_task_states(self):
+        if not self.allow_task_dependencies and self.state == '04_waiting_normal':
+            self.state = '01_in_progress'
```
Mmmh i don't think so,
Here we need to catch the transition of **allow_task_dependencies** from True to False
 -> in this case every task that was in **waiting_normal** is just resets to **in_progress**, How can you make this distinction in the compute_state? isn't the function always going to reset the task to **in_progress** when the **allow_task_dependencies** is False?

xavierbol replied:
I don't think we have to know that we can just check if the `allow_task_dependencies=False` then we do the change accordingly, for instance, if I change a little bit your compute method:
```py
    @api.depends('depend_on_ids.state', 'allow_task_dependencies')
    def _compute_state(self):
        for task in self:
            dependent_open_tasks = []
            if task.allow_task_dependencies:
                dependent_open_tasks = [dependent_task for dependent_task in task.depend_on_ids if dependent_task.state not in CLOSED_STATES]
            # if one of the blocking task is in a blocking state
            if dependent_open_tasks:
                # here we check that the blocked task is not already in a closed state (if the task is already done we don't put it in waiting state)
                if task.state not in CLOSED_STATES:
                    task.state = '04_waiting_normal'
            # if the task as no blocking dependencies and is in waiting_normal, the task goes back to in progress
            elif task.state == '04_waiting_normal':
                task.state = '01_in_progress'
```
when allow_task_dependencies is False we have to do the same than when we don't have any dependencies.

(FYI, `break` stops the for loop, I think here we want to go to the next task in self and so you have to use `continue` instead)

bastvdn replied:
Ok does work indeed, i can remove the `onchange allow_task_dependencies`

bastvdn commented on addons/project/models/project.py:
```diff
@@ -1366,9 +1367,43 @@ def _search_is_private(self, operator, value):
+                        # here we check that the blocked task is not alread in a closed state (if the task is already done we don't put it in waiting state)
+                        if task.state not in CLOSED_STATES:
+                            task.state = '04_waiting_normal'
+                        return
```
If i break there i'll just get out of the last loop and end up in the `if task.state == '04_waiting_normal'` condition. So i changed my code to avoid nested loops by using a list comprehension

bastvdn commented on addons/project/models/project.py:
```diff
@@ -1714,6 +1745,8 @@ def copy(self, default=None):
         if not has_default_name:
             default['name'] = _("%s (copy)", self.name)
+        if self.state:
+            default['state'] = '01_in_progress'
```
:+1: Indeed

bastvdn commented on addons/project/static/src/components/project_task_state_selection/project_task_state_selection.js:
```diff
@@ -0,0 +1,165 @@
+
+const { useState } = owl;
+
+export class ProjectTaskStateSelection extends stateSelectionField.component {
```
:+1:

bastvdn commented on addons/project/static/src/components/project_task_state_selection/project_task_state_selection.js:
```diff
@@ -0,0 +1,165 @@
+        // will determine if a single click will trigger the toggleState() method which will switch the state from in progress to done
+        // Either the isToggleMode is active on the record OR the task is_private
+        return this.props.isToggleMode || this.props.record.data.is_private;
+    }
```
This is actually for the new subtask implementation, when a task has subtasks we can now change those states from the kanban view, using the condition is_private is easier because i don't need to fetch this data for the subtask widget at all (the is_private condition is always false for subtasks anyway).
So this is just easier implementation, tell me if you would prefer i fetch the fields anyway

bastvdn commented on addons/project/static/src/components/project_task_state_selection/project_task_state_selection.js:
```diff
@@ -0,0 +1,165 @@
+            return 'btn btn-link d-flex p-0'
+        }
+        return 'o_state_button btn rounded-pill ' + this.colorButton[currentValue]
+    }
```
:grimacing: sorry forgot about linting

bastvdn commented on addons/project/static/src/components/project_task_state_selection/project_task_state_selection.js:
```diff
@@ -0,0 +1,165 @@
+            for (let word of label.split(" ")) {
+                small_str = small_str + word.slice(0,4) + ". ";
+            }
+            return small_str.substring(0, small_str.length-1);
```
Infact, we don't even need this method anymore ^^'

bastvdn commented on addons/project/static/src/components/project_task_state_selection/project_task_state_selection.xml:
```diff
@@ -0,0 +1,91 @@
+                    <span t-if="showLabel" class="o_status_label ms-1" t-out="label"/>
+                </button>
+            </t>
+        </xpath>
```
[1] for the first index :shrug:

bastvdn commented on addons/project/static/src/components/project_task_state_selection/project_task_state_selection.xml:
```diff
@@ -0,0 +1,91 @@
+        <!-- Waiting state button -->
+        <xpath expr="//t[@t-if='isReadonly']" position="after">
+            <t t-elif="this.props.value == '04_waiting_normal' and ['kanban', 'list'].includes(this.props.record.activeFields[this.props.name].viewType)">
+                <button class="d-flex align-items-center btn fw-normal p-0" style="justify-content: center;" title="This task is blocked by another unfinished task">
```
:+1:

bastvdn commented on addons/project/static/src/components/subtask_kanban_list/subtask_kanban_list.js:
```diff
@@ -35,21 +35,21 @@ export class SubtaskKanbanList extends Component {
                     display_name: { type: "char" },
-                    kanban_state: { selection: [['normal', 'In Progress'], ['done', 'Ready'], ['blocked', 'Blocked']],
+                    state: { selection: [['01_in_progress', 'In Progress'], ['02_changes_requested', 'Changes Requested'], ['03_approved', 'Approved']
+                                ,['04_waiting_normal', 'Waiting'], ['1_done', 'Done'], ['1_canceled', 'Canceled']],
```
:+1:

bastvdn commented on addons/project/tests/test_task_state.py:
```diff
@@ -0,0 +1,142 @@
+        })
+        cls.task_2.write({
+            'project_id': cls.project_goats.id,
+        })
```
:+1:

bastvdn commented on addons/project/tests/test_task_state.py:
```diff
@@ -0,0 +1,142 @@
+        })
+        # 2) change task_1 and task_2 from stage, task_1 should go back to in_progress, task_2 should stay in its closing state
+        (self.task_1 + self.task_2).write({
+            'stage_id': stage_won.id
```
:+1:

bastvdn commented on addons/project/tests/test_task_state.py:
```diff
@@ -0,0 +1,142 @@
+
+    def test_duplicate_dependent_task(self):
+
+        self.task_1.write({
```
:grimacing:

bastvdn commented on addons/project/static/src/components/project_task_state_selection/project_task_state_selection.js:
```diff
@@ -0,0 +1,165 @@
+            "1_canceled": "btn-danger",
+            "04_waiting_normal": "btn-light border-dark",
+        };
+        if (this.props.record.activeFields[this.props.name].viewType !== "form") {
+            super.setup();
+        }
+        else {
+            const hotkeys = ["D", "F", "G", "shift+F", "shift+G"];
+            for (const [index, [value, label]] of this.options.entries()) {
+                useCommand(
+                    sprintf(this.env._t("Set state as %s"), label),
+                    () => this.updateRecord(value),
+                    {
+                        category: "smart_action",
+                        hotkey: "alt+" + hotkeys[index],
+                        isAvailable: () => this.props.value !== value,
+                    }
+                );
+            }
+        }
```
Ok, there as been some last minute change, and now we want a sub-list of hotkeys (like for the **assign to ...** and **Move to Stage ...**) I'll see if i can implement the hook logic

xavierbol commented on addons/project/models/project.py:
```diff
@@ -1366,9 +1367,43 @@ def _search_is_private(self, operator, value):
+        if self.state != '04_waiting_normal' and self.state not in CLOSED_STATES:
+            self.state = '01_in_progress'
+
+    @api.onchange('project_id')
+    def _onchange_project_id(self):
+        if self.state != '04_waiting_normal':
+            self.state = '01_in_progress'
```
you are right, however, I think the compute is triggered when the project changed because we have `allow_task_dependencies` (`project_id.allow_task_dependencies`) in the dependencies, so maybe this onchange is useless. 🤔

xavierbol commented on addons/project/static/src/components/project_task_name_with_subtask_count_char_field/project_task_name_with_subtask_count_char_field.xml:
```diff
@@ -2,6 +2,10 @@
 <templates xml:space="preserve">
 
     <t t-name="project.ProjectTaskNameWithSubtaskCountCharField" t-inherit="web.CharField" t-inherit-mode="primary" owl="1">
+        <xpath expr="//span[@t-esc='formattedValue']" position="replace">
+            <s t-if="['1_done', '1_canceled'].includes(this.props.record.data.state) and ['kanban', 'list'].includes(this.props.record.activeFields[this.props.name].viewType)" t-esc="formattedValue"/>
+            <span t-else="" t-esc="formattedValue"/>
+        </xpath>
```
Why not doing this:
```suggestion
        <xpath expr="//span[@t-esc='formattedValue']" position="before">
            <s t-if="['1_done', '1_canceled'].includes(this.props.record.data.state) and ['kanban', 'list'].includes(this.props.record.activeFields[this.props.name].viewType)" t-esc="formattedValue"/>
        </xpath>
        <xpath expr="//span[@t-esc='formattedValue']" position="attributes">
            <attribute name="t-else">""</attribute>
        </xpath>
```
? 🤔 

What is the `s` element? 🤔 
You don't want to put `span` instead?

xavierbol replied:
oh right 🙂

xavierbol replied:
> Why not doing this:
> 
> ? 🤔
> 
> What is the `s` element? 🤔 You didn't want to put `span` instead?

Don't forget to check that when you have time, I think we can avoid doing a replace here. 🙂 

Thanks in advance!

bastvdn replied:
Yes sorry :) it's done

xavierbol replied:
cool 🤩

xavierbol commented on addons/project/static/src/components/project_task_state_selection/project_task_state_selection.js:
```diff
@@ -0,0 +1,179 @@
+        };
+        if (this.props.record.activeFields[this.props.name].viewType !== "form") {
+            super.setup();
+        }
+        else {
```
```suggestion
        } else {
```

xavierbol reviewed (commented):
Thanks again for your work, I left one comment and I will push a commit if you agree with the changes then you can squash it in yours. 🙂

xavierbol reviewed (approved):
LGTM, thanks again for your work. 🤩 🥳 

May I ask you to edit the PR message to avoid having a big DEPRECATED? Thanks in advance! 🙂 

robodoo r+

xavierbol commented:
robodoo r+
